ctk/common/src/main/kotlin/org/codice/compliance/utils/schema/SchemaValidator.kt

```diff
@@ -0,0 +1,111 @@
+import javax.xml.transform.dom.DOMSource
+import java.io.ByteArrayInputStream
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.stream.StreamResult
+import java.io.ByteArrayOutputStream
+
+class SchemaValidator {
+    companion object {
+
+        private const val JAXP_SCHEMA_LANGUAGE =
+                "http://java.sun.com/xml/jaxp/properties/schemaLanguage"
+        private const val JAXP_SCHEMA_SOURCE =
+                "http://java.sun.com/xml/jaxp/properties/schemaSource"
+        private const val EXTERNAL_SCHEMA_LOCATION =
+                "http://apache.org/xml/properties/schema/external-schemaLocation"
+
+        private val schemas = arrayOf(
+                "saml-schema-assertion-2.0.xsd",
+                "saml-schema-protocol-2.0.xsd"
+        )
+
+        private val schemaLocationOverrides = mapOf(
+                "http://www.w3.org/2000/09/xmldsig#" to "xmldsig-core-schema.xsd",
+                "http://www.w3.org/2001/04/xmlenc#" to "xenc-schema.xsd"
+        )
+
+        fun validateSAMLMessage(saml: Node) {
+            // Convert SAML to InputStream
+            val outputStream = ByteArrayOutputStream()
+            val xmlSource = DOMSource(saml)
+            val outputTarget = StreamResult(outputStream)
+            TransformerFactory.newInstance().newTransformer().transform(xmlSource, outputTarget)
+            val samlInputStream = ByteArrayInputStream(outputStream.toByteArray())
+
+            // Create array of schema InputStreams
+            val schemaInputStreams = schemas
+                    .map { this::class.java.classLoader.getResourceAsStream(it) }
+                    .toTypedArray()
+
+            // Create DocumentBuilder for schema
+            val dbf = DocumentBuilderFactory.newInstance()
+            dbf.isNamespaceAware = true
+            dbf.isValidating = true
+            dbf.setAttribute(JAXP_SCHEMA_LANGUAGE, XMLConstants.W3C_XML_SCHEMA_NS_URI)
+            dbf.setAttribute(JAXP_SCHEMA_SOURCE, schemaInputStreams)
+            dbf.setAttribute(EXTERNAL_SCHEMA_LOCATION, schemaLocationOverrides
+                    .map { it.key to this::class.java.classLoader.getResource(it.value).toExternalForm() }
+                    .map { "${it.first} ${it.second}" }
+                    .joinToString("\n")
+            )
+            dbf.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "file")
+
+            // Validate and report errors
+            val db = dbf.newDocumentBuilder()
+            db.setErrorHandler(SAMLErrorHandler())
+            db.parse(samlInputStream)
+        }
+    }
+
+    private class SAMLErrorHandler : ErrorHandler {
+
+        private fun getParseExceptionInfo(spe: SAXParseException): String {
+            var systemId: String? = spe.systemId
+            if (systemId == null) {
+                systemId = "null"
+            }
+
+            return "URI=" + systemId + " Line=" + spe.lineNumber + ": " + spe.message
```

Comment:
✏️ In Kotlin you can use [String Templates](https://kotlinlang.org/docs/reference/basic-types.html#string-templates) instead of concatenating Strings like this.
✏️ Also you can use the [Elvis Operator](https://kotlinlang.org/docs/reference/null-safety.html#elvis-operator)  to evaluate what `systemId` should be in the null case.

With both of the above, this method could be reduced to one line:
`return "URI=${spe.systemId ?: "null"} Line=${spe.lineNumber}: ${spe.message}"`